qiskit/transpiler/passes/optimization/collect_linear_functions.py

```diff
@@ -13,29 +13,70 @@
 
 """Replace each sequence of CX and SWAP gates by a single LinearFunction gate."""
 
+from collections import deque
 from qiskit.circuit.library.generalized_gates import LinearFunction
 from qiskit.transpiler.basepasses import TransformationPass
 from qiskit.circuit import QuantumCircuit
+from qiskit.dagcircuit import DAGOpNode
+
+
+def collect_linear_blocks(dag):
+    """Collect blocks of linear gates."""
+
+    blocks = []
+
+    in_degree = dict()
+    pending_linear_ops = deque()
+    pending_non_linear_ops = deque()
+
+    def process_node(node):
+        for suc in dag.successors(node):
+            if not isinstance(suc, DAGOpNode):
+                continue
+            in_degree[suc] -= 1
+            if in_degree[suc] > 0:
+                continue
+
+            if suc.name in ["cx", "swap"]:
+                pending_linear_ops.append(suc)
+            else:
+                pending_non_linear_ops.append(suc)
+
+    for node in dag.op_nodes():
+        deg = sum(1 for op in dag.predecessors(node) if isinstance(op, DAGOpNode))
+        if deg > 0:
+            in_degree[node] = deg
+            continue
+
+        if node.name in ["cx", "swap"]:
+            pending_linear_ops.append(node)
+        else:
+            pending_non_linear_ops.append(node)
+
+    while pending_non_linear_ops or pending_linear_ops:
+
+        # first collect as many non linear gates as possible
+        while pending_non_linear_ops:
+            node = pending_non_linear_ops.popleft()
+            process_node(node)
+
+        # now collect as many linear gates as possible
+        cur_block = []
+        while pending_linear_ops:
+            node = pending_linear_ops.popleft()
+            process_node(node)
+            cur_block.append(node)
+
+        if cur_block:
```

Comment:
Again, you were checking that:

```suggestion
        # Collect only blocks comprising at least 2 gates.
        if len(cur_block) > 1:
```